File: apps/web/components/conversations/lentille/lentille-mode-labels.ts
/**
 * `lentille-mode-labels` — WL-108 (LWS-8 « encoche actionnable », LWS-11).
 *
 * Source UNIQUE des noms de mode pour les surfaces web qui les affichent —
 * jumeau exact de `LentilleModeLabels.swift` (I-071/I-072) et sa raison
 * d'être, mot pour mot : « les trois DOIVENT nommer un même mode de la même
 * façon, sinon "AUTO · Focal" sur la carte et "Focale" dans le menu
 * raconteraient deux histoires différentes de la même décision. »
 *
 * Fonctions PURES : elles reçoivent le `t` déjà scopé (`useI18n
 * ('conversations')`), comme `LentilleRow`/`ReadingModeMenu` — jamais
 * d'appel à `useI18n` ici, ce qui les rend testables sans rendu.
 *
 * Les CLÉS ne sont pas neuves (re-prouvé, `apps/web/locales/<locale>/
 * conversations.json`) : `lentille.modes.{auto,focal,script,resume,riviere}` et
 * `lentille.modes.autoBadge` = « AUTO · {decision} » ont été posées par
 * S-005 et sont déjà verrouillées par `__tests__/locales/lentille-i18n-keys
 * .test.ts`. Ce module les CENTRALISE ; `ReadingModeMenu` les épelait en
 * ligne, il les consomme désormais ici (mêmes clés, même rendu).
 *
 * Seule `lentille.modes.bubbles` est ajoutée par WL-108 — le pendant du
 * `lentille.mode.name.bubbles` iOS, pour la même raison DÉFENSIVE : rendre
 * `decisionModeLabel` exhaustif sur `ConversationReadingMode` plutôt que de
 * prétendre un cas inatteignable. `bubbles` est la décision du drapeau
 * ÉTEINT (`resolveOrchestratorDecision`, branche `flag-disabled`) et la
 * carte de focus n'existe que drapeau ALLUMÉ : la clé ne devrait jamais
 * s'afficher, et elle existe quand même.
 *
 * @see tasks/lentille-implementation-contract.md LWS-8
 * @see apps/ios/Meeshy/Features/Main/Lentille/Mode/LentilleModeLabels.swift
 */
import type { ConversationBridge } from '@meeshy/shared/types/conversation-bridge';
import type { ConversationReadingMode, ReadingModePreference } from '@meeshy/shared/types/reading-modes';
import type { OrchestratorDecision } from '@meeshy/shared/utils/reading-modes';
import type { LentilleRowTranslate } from './LentilleRow';

/**
 * Clé i18n du nom d'une PRÉFÉRENCE (mots du menu, chip d'un mode forcé).
 *
 * AMENDEMENT S1 (REV-4bis/B2) — `bulles` RÉUTILISE `lentille.modes.bubbles`,
 * la clé que WL-108 avait déjà posée pour le mode RENDU homonyme (voir
 * `DECISION_LABEL_KEY` ci-dessous, et la note de tête de ce fichier). Aucune
 * clé neuve n'est créée : c'est le même mot pour le lecteur, et `bulles`
 * n'est offert par AUCUN des trois chemins d'entrée du menu de mode
 * (`ReadingModeMenu`, catalogue inchangé) — seul le sélecteur historique
 * `LensSwitcher` l'écrit, drapeau éteint. La table reste TOTALE sur
 * l'énumération amendée, ce qui est sa seule raison d'exister :
 * `preferenceLabel` ne doit jamais rendre `undefined`.
 */
const PREFERENCE_LABEL_KEY: Readonly<Record<ReadingModePreference, string>> = {
  auto: 'lentille.modes.auto',
  focal: 'lentille.modes.focal',
  script: 'lentille.modes.script',
  resume: 'lentille.modes.resume',
  riviere: 'lentille.modes.riviere',
  bulles: 'lentille.modes.bubbles',
};

/**
 * Clé i18n du nom d'un mode RENDU (`OrchestratorDecision.mode`). Le
 * catalogue des modes rendus et celui des préférences ne coïncident pas :
 * `auto` n'est pas un mode (il rend la main à la loi), `summary`/`river`
 * portent les noms `resume`/`riviere` côté menu, et `bubbles` n'a pas de
 * préférence correspondante.
 */
const DECISION_LABEL_KEY: Readonly<Record<ConversationReadingMode, string>> = {
  focal: 'lentille.modes.focal',
  script: 'lentille.modes.script',
  summary: 'lentille.modes.resume',
  river: 'lentille.modes.riviere',
  bubbles: 'lentille.modes.bubbles',
};

/** Nom d'une préférence — le libellé du menu ET du chip d'un mode forcé. */
export function preferenceLabel(preference: ReadingModePreference, t: LentilleRowTranslate): string {
  return t(PREFERENCE_LABEL_KEY[preference]);
}

/** Nom du mode que l'orchestrateur a RÉELLEMENT rendu — utilisé dans « AUTO · … » seulement. */
export function decisionModeLabel(mode: ConversationReadingMode, t: LentilleRowTranslate): string {
  return t(DECISION_LABEL_KEY[mode]);
}

/**
 * R6-5 — projection INVERSE de `toBridgeSuggestedMode`
 * (`packages/shared/utils/reading-modes.ts`) : `bridge.suggestedMode` n'est
 * PAS une nouvelle loi, c'est déjà un mode RENDU légal (le sous-ensemble
 * `'focal' | 'resume'` que la loi projette), reçu tel quel du wire. Cette
 * fonction ne fait que le reconvertir vers le vocabulaire de
 * `decisionModeLabel` (`ConversationReadingMode`) pour partager EXACTEMENT
 * les mêmes clés i18n que le recalcul local — jamais un second nommage.
 * `'resume'` → `'summary'` (même mot, deux catalogues voisins,
 * cf. `ConversationReadingMode` vs `ConversationBridge.suggestedMode`) ;
 * `'focal'` → `'focal'`, identité.
 */
function bridgeSuggestedRenderMode(suggestedMode: ConversationBridge['suggestedMode']): ConversationReadingMode {
  return suggestedMode === 'resume' ? 'summary' : 'focal';
}

/**
 * Texte de l'encoche (critère LWS-8 : « elle affiche "AUTO · <décision
 * courante>" — l'utilisateur voit ce qui VA se passer, pas une étiquette
 * générique »).
 *
 * Préférence `auto` ⇒ prévision « AUTO · <décision> ». Mode mémorisé ⇒ le
 * nom du mode SEUL : ce n'est plus une prévision mais un CHIP, l'orchestrateur
 * étant débrayé pour cette conversation.
 *
 * R6-5 — HIÉRARCHIE SERVEUR PUIS LOCAL, jamais l'inverse : `suggestedMode`
 * (`conversation.bridge?.suggestedMode`, précalculé par les 3 producteurs —
 * gateway `ConversationBridgeService`, substitut `LocalBridgeProvider`, iOS
 * `LentilleProviders` — via la MÊME loi que `decision`) PRIME quand il est
 * PRÉSENT. `decision.mode` (le recalcul LOCAL de ce composant,
 * `resolveOrchestratorDecision` rejoué sur les données déjà en main) reste le
 * SEUL repli quand `suggestedMode` est ABSENT (aucun pont sur cette
 * conversation) — jamais un vide, jamais une valeur inventée : c'est
 * exactement ce que `resolveOrchestratorDecision` aurait rendu ici de toute
 * façon, avant ce branchement.
 *
 * Q142-c (2026-08-18, TRANCHÉE) — LES DEUX PRIMENT SOUS `isReadingModesFlagActive`.
 * Ni `suggestedMode` ni `decision.mode` ne connaissent le défaut provisoire
 * « Bulles » du fil WEB (`PROVISIONAL_DEFAULT_RENDER`,
 * `hooks/lentille/use-thread-reading-mode.ts`) : c'est une décision
 * exclusivement web, jamais amendée dans la loi partagée que gateway/iOS/le
 * recalcul local appellent tous les trois. Sans choix explicite du lecteur
 * (`preference === 'auto'`) ET le drapeau du fil (`useReadingModesFlag`) allumé, OUVRIR cette
 * conversation dans le fil web rend les bulles — quoi que `suggestedMode` ou
 * `decision.mode` aient prédit. `isReadingModesFlagActive` (résolu par
 * l'appelant, `LentillePeek`, jamais recalculé ici) fait donc primer «
 * Bulles » sur les deux avant même de les regarder : DÉFAUT à `false`, donc
 * AUCUN appelant existant n'est affecté tant qu'il ne fournit pas ce
 * cinquième argument (non-régression R6-4/R6-5).
 *
 * @see hooks/lentille/use-thread-reading-mode.ts (`useThreadActiveReadingMode`)
 *      — même correction, pour `LensSwitcher`.
 */
export function notchText(
  decision: OrchestratorDecision,
  preference: ReadingModePreference,
  t: LentilleRowTranslate,
  suggestedMode?: ConversationBridge['suggestedMode'] | null,
  isReadingModesFlagActive = false
): string {
  if (preference !== 'auto') return preferenceLabel(preference, t);
  if (isReadingModesFlagActive) {
    return t('lentille.modes.autoBadge', { decision: decisionModeLabel('bubbles', t) });
  }
  const mode = suggestedMode != null ? bridgeSuggestedRenderMode(suggestedMode) : decision.mode;
  return t('lentille.modes.autoBadge', { decision: decisionModeLabel(mode, t) });
}

File: apps/ios/Meeshy/Features/Main/Lentille/Mode/LentilleModeLabels.swift
import Foundation

/// Libellés des modes de lecture — source UNIQUE pour les trois surfaces qui
/// les affichent (contrat LWS-8) : l'encoche/chip de la focus card (I-071),
/// le sous-menu « Mode de lecture » et l'aperçu (I-072). Les trois DOIVENT
/// nommer un même mode de la même façon, sinon « AUTO · Focal » sur la carte
/// et « Focale » dans le menu raconteraient deux histoires différentes de la
/// même décision.
///
/// Pur, `nonisolated`, testable sans vue. Toute chaîne passe par
/// `String(localized:defaultValue:bundle:)` — même convention que le reste de
/// `Lentille/**` (`LentilleBridgeLine`, `LentilleSectionIdentity`) : Xcode
/// extrait la clé au build, la valeur par défaut s'affiche immédiatement dans
/// toute langue non encore traduite.
///
/// i18n — ré-preuve faite (mission I-072, §0) : AUCUNE clé de mode de lecture
/// n'existe côté iOS (`apps/ios/Meeshy/Localizable.xcstrings`, recherche
/// `reading.?mode|lentille\.mode|context\.mode` — zéro résultat). Les clés
/// S-005 du contrat vivent côté WEB (`apps/web/locales/*/conversations.json`)
/// et ne sont pas réutilisables ici (catalogues iOS/web disjoints). Ce
/// fichier introduit donc les clés `lentille.mode.*`, dans la continuité
/// directe des `lentille.bridge.*` / `lentille.section.*` déjà posées par
/// I-065/I-066 sur ce même modèle — PAS une invention isolée, la poursuite
/// d'un précédent déjà posé trois fois dans ce dossier.
///
/// @see tasks/lentille-implementation-contract.md LWS-8
/// @see tasks/lentille-workshop-execution.md I-071, I-072
nonisolated enum LentilleModeLabels {

    // MARK: - Nom du mode — préférence (menu, chip forcé)

    /// Titre affiché dans le menu et sur le chip forcé (encoche quand
    /// `preference != .auto`). `.auto` porte son propre libellé ("Auto"),
    /// distinct du nom de mode qu'il RÉSOUT (cf. `decisionModeTitle`).
    static func menuTitle(for preference: ReadingModeOrchestrator.ReadingModePreference) -> String {
        switch preference {
        case .auto:
            return String(localized: "lentille.mode.name.auto", defaultValue: "Auto", bundle: .main)
        case .focal:
            return String(localized: "lentille.mode.name.focal", defaultValue: "Focal", bundle: .main)
        case .script:
            return String(localized: "lentille.mode.name.script", defaultValue: "Script", bundle: .main)
        case .resume:
            return String(localized: "lentille.mode.name.resume", defaultValue: "Résumé", bundle: .main)
        case .riviere:
            return String(localized: "lentille.mode.name.riviere", defaultValue: "Rivière", bundle: .main)
        // AMENDEMENT S1 (REV-4bis/B2) — jamais affiché sur iOS : `.bulles` est
        // hors de l'ordre du menu (`LentilleModeMenu.build`, cinq entrées).
        // RÉUTILISE la clé du mode RENDU `.bubbles` (cf. `decisionModeTitle`
        // ci-dessous) plutôt que d'en créer une sixième : c'est le même mot
        // pour le lecteur, et une clé i18n neuve pour un libellé jamais rendu
        // serait une dette gratuite dans les 12 catalogues de localisation.
        case .bulles:
            return String(localized: "lentille.mode.name.bubbles", defaultValue: "Bulles", bundle: .main)
        }
    }

    // MARK: - Nom du mode — décision RENDUE (encoche « AUTO · <décision> »)

    /// Titre du mode que l'orchestrateur a RÉELLEMENT rendu
    /// (`OrchestratorDecision.mode`) — utilisé uniquement dans « AUTO · … ».
    /// `.bubbles` (drapeau désactivé) n'apparaît jamais dans la carte (elle
    /// n'existe que drapeau actif) ; le libellé défensif évite un `switch`
    /// non-exhaustif plutôt qu'un cas qu'on prétendrait inatteignable.
    static func decisionModeTitle(for mode: ReadingModeOrchestrator.ConversationReadingMode) -> String {
        switch mode {
        case .focal:
            return String(localized: "lentille.mode.name.focal", defaultValue: "Focal", bundle: .main)
        case .script:
            return String(localized: "lentille.mode.name.script", defaultValue: "Script", bundle: .main)
        case .summary:
            return String(localized: "lentille.mode.name.resume", defaultValue: "Résumé", bundle: .main)
        case .river:
            return String(localized: "lentille.mode.name.riviere", defaultValue: "Rivière", bundle: .main)
        case .bubbles:
            return String(localized: "lentille.mode.name.bubbles", defaultValue: "Bulles", bundle: .main)
        }
    }

    // MARK: - Texte de l'encoche

    /// « AUTO · <décision> » quand la préférence est `.auto` (l'utilisateur
    /// voit ce qui VA se passer, contrat LWS-8) ; le nom du mode forcé SEUL
    /// quand un mode est mémorisé (M-048) — c'est alors un CHIP, pas une
    /// prévision.
    static func notchText(
        decision: ReadingModeOrchestrator.OrchestratorDecision,
        preference: ReadingModeOrchestrator.ReadingModePreference
    ) -> String {
        guard preference == .auto else {
            return menuTitle(for: preference)
        }
        let format = String(localized: "lentille.mode.notch.auto", defaultValue: "AUTO · %@", bundle: .main)
        return String(format: format, decisionModeTitle(for: decision.mode))
    }

    // MARK: - Raison Rivière — seuils VIVANTS, jamais un texte fixe

    /// Trois formes, jamais une seule formule (AMENDEMENT S1, REV-3/B3) —
    /// composées depuis `RiverEligibilityReason` (miroir gelé), jamais une
    /// chaîne statique : `threshold`, `current` et `riverReason` viennent de
    /// `resolveCapabilities`, appelée avec les données RÉELLES de la
    /// conversation à chaque rendu.
    ///
    /// 1. `.neverEligible` (conversation `direct`) ⇒ « jamais en conversation
    ///    directe ». L'ancienne formule unique promettait « s'ouvrira à 5
    ///    personnes actives — N aujourd'hui » à un duo qui n'atteindra JAMAIS
    ///    5 : une porte annoncée qui n'existe pas.
    /// 2. Compte INCONNU (`current == nil`) ⇒ le seuil SEUL, « s'ouvrira à 5
    ///    personnes actives ». Aucun « 0 aujourd'hui » fabriqué : le compte
    ///    d'actifs par conversation n'est pas encore une donnée client (G-123).
    /// 3. Compte connu ⇒ la formule à deux nombres, INCHANGÉE.
    ///
    /// `.eligible` retombe sur la même branche numérique que
    /// `.belowThreshold` : en V3 l'entrée Rivière reste grisée même jugée
    /// éligible (drapeau `riviere_mode` absent, R-133), et sa raison affichée
    /// est alors la même formule vivante qu'avant l'amendement.
    static func riverReason(_ reason: ReadingModeOrchestrator.RiverEligibilityReason) -> String {
        if reason.riverReason == .neverEligible {
            return String(
                localized: "lentille.mode.river.never",
                defaultValue: "jamais en conversation directe",
                bundle: .main
            )
        }

        guard let current = reason.current else {
            let thresholdOnly = String(
                localized: "lentille.mode.river.threshold_only",
                defaultValue: "s'ouvrira à %d personnes actives",
                bundle: .main
            )
            return String(format: thresholdOnly, reason.threshold)
        }

        let format = String(
            localized: "lentille.mode.river.reason",
            defaultValue: "s'ouvrira à %d personnes actives — %d aujourd'hui",
            bundle: .main
        )
        return String(format: format, reason.threshold, current)
    }
}
